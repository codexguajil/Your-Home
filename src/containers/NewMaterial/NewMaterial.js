import React from 'react';
import { Button, Modal, Form, Input } from 'antd';

const CollectionCreateForm = Form.create({ name: 'form_in_modal' })(
  class extends React.Component {
    render() {
      const { visible, onCancel, onCreate, form } = this.props;
      const { getFieldDecorator } = form;
      return (
        <Modal
          visible={visible}
          title="Add a Material"
          okText="Create"
          onCancel={onCancel}
          onOk={onCreate}
        >
          <Form layout="vertical">
            <Form.Item label="Name">
              {getFieldDecorator('name', {
                rules: [{ required: true, message: 'Please input the name of the material!' }],
              })(<Input />)}
            </Form.Item>
            <Form.Item label="Brand">
              {getFieldDecorator('brand', {
                rules: [{ required: true, message: 'Please input the brand of the material!' }],
              })(<Input />)}
            </Form.Item>
            <Form.Item label="Model">
              {getFieldDecorator('model', {
                rules: [{ required: true, message: 'Please input the model number of the material!' }],
              })(<Input />)}
            </Form.Item>
            <Form.Item label="Vendor">
              {getFieldDecorator('vendor', {
                rules: [{ required: true, message: 'Please input where the material was purchased!' }],
              })(<Input />)}
            </Form.Item>
            <Form.Item label="Quantity">
              {getFieldDecorator('quantity', {
                rules: [{ required: true, message: 'Please input how many of this item was purchased!' }],
              })(<Input />)}
            </Form.Item>
            <Form.Item label="Price">
              {getFieldDecorator('price', {
                rules: [{ required: true, message: 'Please input the price of the material!' }],
              })(<Input />)}
            </Form.Item>
            <Form.Item label="Manual URL">
              {getFieldDecorator('manual', {
                rules: [{ required: false, message: 'Please input the price of the material!' }],
              })(<Input />)}
            </Form.Item>
            <Form.Item label="Notes">
              {getFieldDecorator('notes')(<Input type="textarea" />)}
            </Form.Item>
            <Form.Item className="collection-create-form_last-form-item">
            </Form.Item>
          </Form>
        </Modal>
      );
    }
  }
);

class NewMaterial extends React.Component {
  state = { visible: false };

  showModal = () => {
    this.setState({ visible: true });
  }

  handleCancel = () => {
    this.setState({ visible: false });
  }

  handleCreate = () => {
    const form = this.formRef.props.form;
    form.validateFields((err, values) => {
      if (err) {
        return;
      }

      console.log('Received values of form: ', values);
      form.resetFields();
      this.setState({ visible: false });
    });
  }

  saveFormRef = formRef => {
    this.formRef = formRef;
  }

  render() {
    return (
      <div>
        <Button type="primary" onClick={this.showModal}>
          Add New Material
        </Button>
        <CollectionCreateForm
          wrappedComponentRef={this.saveFormRef}
          visible={this.state.visible}
          onCancel={this.handleCancel}
          onCreate={this.handleCreate}
        />
      </div>
    );
  }
}

export default NewMaterial;