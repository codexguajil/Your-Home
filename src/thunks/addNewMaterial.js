import { isLoading, hasErrored, addMaterial } from '../actions';

export const addNewMaterial = (body) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify(body),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      console.log(data)
      dispatch(isLoading(false))
      // dispatch(addMaterial(data.data.createProject))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}