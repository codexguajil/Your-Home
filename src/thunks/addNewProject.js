
import { isLoading, hasErrored, addProject } from '../actions';

export const addNewProject = (body) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify(body),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      dispatch(isLoading(false))
      dispatch(addProject(data.data.createProject))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}
