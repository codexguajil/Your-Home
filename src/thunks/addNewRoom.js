import { isLoading, hasErrored, addRoom } from '../actions';

export const addNewRoom = (id, name, description, type) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify({"query": `mutation { addProjectRoom (project_id: "${id}", room: {name: "${name}", description: "${description}", type: "${type}"}) {id name description type project {id}}}`}),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json();
      console.log(data.data.addProjectRoom);
      dispatch(isLoading(false));
      dispatch(addRoom(data.data.addProjectRoom));
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}