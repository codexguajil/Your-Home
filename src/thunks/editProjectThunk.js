// { "query":"mutation{updateProject(project: {id: 2, name: \"Updated House\", city: \"Different City\"}) {id name city state}}"
// }

// name, address, description

import { isLoading, hasErrored, editProject } from '../actions';

export const editProjectThunk = (id, name, address, description) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify({
          "query": `mutation {updateProject (project: {id: "${id}", name: "${name}", address: "${address}", description: "${description}"}) {id name address description}}`
          }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      console.log(data.data.updateProject)
      dispatch(isLoading(false))
      dispatch(editProject(data.data.updateProject))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}