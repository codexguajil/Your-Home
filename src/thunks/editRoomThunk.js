import { isLoading, hasErrored } from '../actions';

export const editRoomThunk = (id, name, type, description) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify({
          "query": `mutation {updateRoom (room: {id: "${id}", name: "${name}", type: "${type}", description: "${description}"}) {id name type description roomMaterials{id}}}`
          }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      await response.json()
      dispatch(isLoading(false))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}