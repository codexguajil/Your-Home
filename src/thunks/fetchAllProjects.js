import { isLoading, hasErrored, setProjects, setMaterials } from '../actions';

export const fetchAllProjects = () => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify({
          "query": "query { projects { id name description address rooms { name type description }}}"
          }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      dispatch(isLoading(false))
      dispatch(setProjects(data.data.projects))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}

export const fetchRoomMaterials = () => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify({
          "query": "query { getRoomsMaterials(room_id: 1)}"
          }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      console.log(data.data.getRoomsMaterials)
      dispatch(isLoading(false))
      dispatch(setMaterials(data.data.getRoomsMaterials))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}
