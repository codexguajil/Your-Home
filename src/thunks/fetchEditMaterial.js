import { isLoading, hasErrored, editMaterial } from '../actions';

export const fetchEditMaterial = (request) => {
  return async (dispatch) => {
    console.log(request.body)

    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify(request.body),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      console.log(data)
      dispatch(isLoading(false))
      dispatch(editMaterial(data.data.updateProjectMaterial, request.oldType))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}