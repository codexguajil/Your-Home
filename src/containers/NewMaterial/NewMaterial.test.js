import React from 'react';
import { shallow } from 'enzyme';
import { NewMaterial, mapDispatchToProps } from './NewMaterial';
import { addNewMaterial } from '../../thunks/addNewMaterial';

jest.mock('../../thunks/addNewMaterial');

describe('NewMaterial', () => {
  let wrapper;
  let mockFn;
  beforeEach(() => {
    mockFn = jest.fn();
    wrapper = shallow(<NewMaterial addNewMaterial={mockFn} forceRender={mockFn}/>)
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state visible when showModal is called', () => {
    wrapper.setState({ visible: false })
    wrapper.instance().showModal();
    expect(wrapper.state('visible')).toBe(true);
  });

  it('should updateState for correct name and value on onChange', () => {
    const mockedEvent = { target: {name: 'name', value: 'new name'} };
    wrapper.instance().onChange(mockedEvent);
    expect(wrapper.state('name')).toBe('new name');
  });

  it('should update state visible when handleCancel is called', () => {
    wrapper.setState({ visible: true })
    wrapper.instance().handleCancel();
    expect(wrapper.state('visible')).toBe(false);
  });
  
  it('should update state visible on handleCreate', async () => {
    wrapper.setState({ visible: true });
    await wrapper.instance().handleCreate();
    expect(wrapper.state('visible')).toBe(false);
  });

  it('should dispatch addNewMaterial', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = addNewMaterial();

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addNewMaterial();

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

});