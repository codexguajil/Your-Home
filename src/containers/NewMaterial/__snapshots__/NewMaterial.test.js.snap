// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewMaterial should match snapshot 1`] = `
<div>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[Function]}
    type="primary"
  >
    Add New Material
  </Button>
  <Form(Component)
    onCancel={[Function]}
    onCreate={[Function]}
    visible={false}
    wrappedComponentRef={[Function]}
  />
</div>
`;
